### PR TITLE
Avoid recompiling shaders when possible

### PR DESCRIPTION
**Pull request by shoaibkamil**
My recent change (fixed by @abadams) to allow multiple GLSL stages made
it so that the runtime recompiles shaders even when not necessary.
This change checks (by kernel name) to see if the kernel(s) have
already been compiled and avoids the overhead of recompilation.

We _might_ want to do something more sophisticated for matching
(perhaps hashing the kernel text or something).

This fixes the slowdown reported by Lee Yuguang.

### CODE REVIEW

File: src/runtime/opengl.cpp [hunk 1]
```diff
@@ -1930,6 +1930,28 @@ WEAK int halide_opengl_initialize_kernels(void *user_context, void **state_ptr,
...
     ModuleState *module = *state;
 
+    if (module) {
+        // Check to see if we've already compiled this set of kernels
+        // Currently, this check occurs by name.  We assume that if
+        // the *first* kernel was already compiled, then all of them
+        // were.
+
+        debug(user_context) << "Module already exists.\n";
+        const char* kernel_name = match_prefix(src, kernel_marker);
```
Comment (abadams) [inline src/runtime/opengl.cpp]: How unique are kernel names? What if you have multiple Halide pipelines with different opengl stages that use the same Func name?

**Comment by shoaibkamil** [comment]
Actually, lets hold off on this.  I think the right approach is going to have to do something smarter than match by name.  Perhaps a fast hash of the text?  I'd appreciate any suggestions.

**Comment by zvookin** [comment]
The interface between the GPU API codegen support and the runtime is not
fixed. It would seem prudent to change that API to a) pass the many shaders
as separate items and to b) pass a hash or other unique identifier that is
computed at compile time if needed. Though in practice I think the setup
gives a module handle back which can be used to hold onto state in such a
way that index by hash mechanisms are not needed. I'll try to take a look
at the code in a little bit to give specifics.

File: src/runtime/opengl.cpp (hunk 1 above)
Comment (shoaibkamil) [inline src/runtime/opengl.cpp]: Yeah, that's why I asked for advice on this.  Using kernel names for matching is a bad thing.

**Comment by shoaibkamil** [comment]
@zvookin In the other GPU backends, it's possible to have a single `initialize_kernels()` call that builds (e.g. Metal) or loads (e.g. ptx) a library, and then easily call into the library.  For OpenGL, we have to define/call `main()` for some shader program (I'm sure you know this, but that seems to be the meaningful difference).

In this case, I *think* the right thing to do is to generate a different `initialize_kernels()` call for each stage/func that's going to be compiled to GLSL; that way we can have a module-per-stage/func, and since the `initialize_kernels()` call returns a module pointer that's used in the `run()` call, we wouldn't have to do a run-time lookup.

Is this the design you were thinking about?

**Comment by shoaibkamil** [comment]
Closing this while I implement a better version.

**Pull request closed without merging**